packages/node-cli/src/core/sdksetup/ProxyAgent.js

```diff
@@ -0,0 +1,92 @@
+/*
+ ** Copyright (c) 2021 Oracle and/or its affiliates.  All rights reserved.
+ ** Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+ */
+const http = require('http');
+const { URL } = require('url');
+const tls = require('tls');
+const net = require('net');
+
+class ProxyAgent {
+	constructor(proxyString, options) {
+		this.proxyString = proxyString;
+		this.options = options;
+		this.initProxyURL();
+	}
+
+	initProxyURL() {
+		// throws error if proxyString is invalid
+		const proxy = new URL(this.proxyString);
+
+		proxy.host = proxy.hostname || proxy.host;
+		proxy.port = +proxy.port || (proxy.protocol.toLowerCase() === 'https:' ? 443 : 80);
+		this.proxyURL = proxy;
+	}
+
+	addRequest(req, options) {
+		req.shouldKeepAlive = false;
+
+		this.createConnection(options)
+			.then((socket) => {
+				req.onSocket(socket);
+			})
+			.catch((err) => {
+				req.emit('error', err);
+			});
+	}
+
+	createConnection(options) {
+		return new Promise((resolve, reject) => {
+			const ssl = options.protocol ? options.protocol.toLowerCase() === 'https:' : false;
+			if (ssl && this.options.tunnel === true) {
+				if (options.port === 80) options.port = 443;
+				// CONNECT Method
+				const req = http.request({
+					host: this.proxyURL.hostname,
+					port: this.proxyURL.port,
+					auth: this.proxyURL.auth,
+					method: 'CONNECT',
+					path: (options.hostname || options.host) + ':' + options.port,
+					headers: {
+						host: options.host,
+					},
+					timeout: this.options.timeout,
+				});
+
+				req.once('connect', (res, socket, head) => {
+					const tunnel = tls.connect({
+						socket: socket,
+						host: options.hostname || options.host,
+						port: +options.port,
+						servername: options.servername || options.host,
+					});
+					resolve(tunnel);
+				});
+
+				req.once('timeout', () => {
+					req.destroy();
+					reject(new Error('ProxyAgent CONNECT request timeout.'));
+				});
+
+				req.once('error', (err) => {
+					reject(err);
+				});
+
+				req.once('close', () => {
+					reject(new Error('Tunnel failed. Socket closed prematurely'));
```

Comment:
Move to message and get review from tech writer

Reply:
Reviewed.